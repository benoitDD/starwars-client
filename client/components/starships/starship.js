import React, {Component} from 'react'
import {Query} from 'react-apollo'
import {Starship as StarshipQuery} from '../../query/starships.gql'
import PropTypes from 'prop-types'
import Properties from '../utils/properties'
import {Pilots} from '../utils/linkObjects'
import Images from '../utils/images'
import AddImageHeader from '../utils/addImageHeader'
import '../objects.sass'
import Loading from '../utils/loading'
import HandleError from '../utils/handleError'

function getProperties(starship){
	return [
		{
			label: 'Model',
			value: starship.model
		},
		{
			label: 'Starship class',
			value: starship.starshipClass
		},
		{
			label: 'Manufacturers',
			value: starship.manufacturers
		},
		{
			label: 'Cost in credits',
			value: starship.costInCredits,
			unit: 'credits'
		},
		{
			label: 'Length',
			value: starship.length,
			unit: 'm'
		},
		{
			label: 'Crew',
			value: starship.crew,
			helper: 'The number of personnel needed to run or pilot this starship'
		},
		{
			label: 'Passengers',
			value: starship.passengers,
			helper: 'The number of non-essentiel people this starship can transport'
		},
		{
			label: 'Max atmosphering speed',
			value: starship.maxAtmospheringSpeed,
			helper: 'The maximum speed of this starship in atmosphere'
		},
		{
			label: 'Hyperdrive rating',
			value: starship.hyperdriveRating,
			helper: 'The class of this starships hyperdrive'
		},
		{
			label: 'MGLT',
			value: starship.MGLT,
			helper: 'The Maximum number of Megalights this starship can travel in a standard hour'
		},
		{
			label: 'cargo capacity',
			value: starship.cargoCapacity,
			unit: 'kg'
		},
		{
			label: 'Consumables',
			value: starship.consumables,
			helper: 'The maximum length of time that this starship can provide consumables for its entire crew without having to resupply'
		}
	]
}

class Starship extends Component {
	render(){
		return (
			<Query query = {StarshipQuery} variables = {{id:this.props.id}} fetchPolicy = "network-only">
				{
					({data, loading, error}) => {
						if (loading) return <div className = 'object_loading'><Loading/></div>
						if (error) return <HandleError error = {error}/>
						const {starship} = data
						const properties = getProperties(starship)
						return (
							<section>
								<h1>{starship.name}</h1>
								<article>
									<Images images = {starship.more && starship.more.imagesHeader} idOwn = {starship.id} />
									<AddImageHeader id = {starship.id} typeObject = 'Starship' />
									<div className = 'object_descriptions'>
										<Properties properties = {properties}/>
										<Pilots pilots = {starship.persons ? starship.persons.persons : undefined} />
									</div>
								</article>
							</section>
						)
					}
				}
			</Query>
		)
	}
}

Starship.propTypes = {
	id: PropTypes.string,
	path: PropTypes.string
}

export default Starship