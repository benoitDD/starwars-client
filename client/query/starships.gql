#import './fragments.gql'

query AllStarships($pageSize: Int, $after: String, $before: String){
  allStarships(pageSize: $pageSize, after: $after, before: $before) {
    pageInfo {
      endCursor
      startCursor
      hasNextPage
      hasPreviousPage
      totalCount
    }
    starships {
      name
      id
    }
  }
}

query Starship($id: ID!){
  starship(id: $id) {
    ...StarshipFragmentMore
  }
}

fragment StarshipFragment on Starship {
  id
  name
  model
  starshipClass
  manufacturers
  costInCredits
  length
  crew
  passengers
  maxAtmospheringSpeed
  hyperdriveRating
  MGLT
  cargoCapacity
  consumables
}