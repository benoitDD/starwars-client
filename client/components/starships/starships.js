import React, {Component} from 'react'
import {LinkStarship} from '../utils/links'
import {AllStarships} from '../../query/starships.gql'
import Pagination from '../utils/pagination'

class Starships extends Component {
	render(){
		return (
			<Pagination query = {AllStarships} nameQuery = 'allStarships' nameObjects = 'starships'>
				{
					starship => {
						return (
							<LinkStarship id = {starship.id}>
								{starship.name}
							</LinkStarship>
						)
					}
				}
			</Pagination>
		)
	}
}

export default Starships