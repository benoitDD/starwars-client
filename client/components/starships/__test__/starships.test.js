import React from 'react'
import renderer from 'react-test-renderer'
import {MockedProvider} from 'react-apollo/test-utils'
import wait from 'waait'
import {AllStarships} from '../../../query/starships.gql'
import Starships from '../starships'
import dataStarships from '../../test/starships.json'

const mocks = [
	{
		request: {
			query: AllStarships,
			variables: {
				pageSize: 10
			}
		},
		result: dataStarships
	}
]

test('Test starhips',async (done) => {
	const component = renderer.create(
		<MockedProvider mocks={mocks}>
			<Starships/>
		</MockedProvider>,
	)

	await wait(0)
    
	let tree = component.toJSON()
	expect(tree).toMatchSnapshot()
	done()
})

test('Test starhips loading',async (done) => {
	const component = renderer.create(
		<MockedProvider mocks={mocks}>
			<Starships/>
		</MockedProvider>,
	)
    
	let tree = component.toJSON()
	expect(tree).toMatchSnapshot()
	done()
})

test('Test starhips error',async (done) => {
	const mocksError = [
		{
			request: {
				query: AllStarships,
				variables: {
					pageSize: 10
				}
			},
			result: dataStarships,
			error: new Error('Une erreur :-)')
		}
	]

	const component = renderer.create(
		<MockedProvider mocks={mocksError}>
			<Starships/>
		</MockedProvider>,
	)

	await wait(0)
    
	let tree = component.toJSON()
	expect(tree).toMatchSnapshot()
	done()
})