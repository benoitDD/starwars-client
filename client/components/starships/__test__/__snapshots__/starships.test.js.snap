// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test starhips 1`] = `
<div
  className="pagination"
>
  <div
    className="pagination-header"
  >
    <span>
      37
       
      item.s
    </span>
    <span
      className="pagination-actions"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        previous
      </button>
      <button
        disabled={false}
        onClick={[Function]}
      >
        next
      </button>
    </span>
  </div>
  <ul
    className="pagination-list"
  >
    <li
      className="pagination-list-item"
    >
      <a
        className="link"
        href="/starship/c3RhcnNoaXBzOjQx"
        onClick={[Function]}
      >
        Scimitar
      </a>
    </li>
    <li
      className="pagination-list-item"
    >
      <a
        className="link"
        href="/starship/c3RhcnNoaXBzOjQz"
        onClick={[Function]}
      >
        J-type diplomatic barge
      </a>
    </li>
    <li
      className="pagination-list-item"
    >
      <a
        className="link"
        href="/starship/c3RhcnNoaXBzOjQ3"
        onClick={[Function]}
      >
        AA-9 Coruscant freighter
      </a>
    </li>
    <li
      className="pagination-list-item"
    >
      <a
        className="link"
        href="/starship/c3RhcnNoaXBzOjQ4"
        onClick={[Function]}
      >
        Jedi starfighter
      </a>
    </li>
    <li
      className="pagination-list-item"
    >
      <a
        className="link"
        href="/starship/c3RhcnNoaXBzOjQ5"
        onClick={[Function]}
      >
        H-type Nubian yacht
      </a>
    </li>
    <li
      className="pagination-list-item"
    >
      <a
        className="link"
        href="/starship/c3RhcnNoaXBzOjUy"
        onClick={[Function]}
      >
        Republic Assault ship
      </a>
    </li>
    <li
      className="pagination-list-item"
    >
      <a
        className="link"
        href="/starship/c3RhcnNoaXBzOjU4"
        onClick={[Function]}
      >
        Solar Sailer
      </a>
    </li>
    <li
      className="pagination-list-item"
    >
      <a
        className="link"
        href="/starship/c3RhcnNoaXBzOjU5"
        onClick={[Function]}
      >
        Trade Federation cruiser
      </a>
    </li>
    <li
      className="pagination-list-item"
    >
      <a
        className="link"
        href="/starship/c3RhcnNoaXBzOjYx"
        onClick={[Function]}
      >
        Theta-class T-2c shuttle
      </a>
    </li>
    <li
      className="pagination-list-item"
    >
      <a
        className="link"
        href="/starship/c3RhcnNoaXBzOjYz"
        onClick={[Function]}
      >
        Republic attack cruiser
      </a>
    </li>
  </ul>
</div>
`;

exports[`Test starhips error 1`] = `
<div>
  Une erreur inconnue s'est produite 
  : Network error: Une erreur :-)
</div>
`;

exports[`Test starhips loading 1`] = `
<div
  id="pagination-loading"
/>
`;
